### PR TITLE
Consider lightweight deleted rows when selecting parts to merge

### LINKED ISSUE #56728: Merge issue about large data part with lightweight delete
**Use case**

When a data part is larger than `max_bytes_to_merge_at_max_space_in_pool`, lightweight deleted data inside the data part will never be really deleted unless `OPTIMIZE FINAL` is executed.

**Describe the solution you'd like**

When selecting parts to merge, use estimated existed rows size
 (`bytes_on_disk * (rows_count - lwd_rows_count) / rows_count`) as the source part size instead of `bytes_on_disk`.
We can enable this mode only when deleted rows ratio exceeds a certain value (maybe 25%).

### PR DESCRIPTION
**Pull request by jewelzqiu**
resolves #56728 

### Changelog category:
- Improvement

### Changelog entry:
Consider lightweight deleted rows when selecting parts to merge if enabled

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 76bec6445e2a2c24591af0cfc092403f2fc8deb8 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57648/76bec6445e2a2c24591af0cfc092403f2fc8deb8/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7138683247">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57648/76bec6445e2a2c24591af0cfc092403f2fc8deb8/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by jewelzqiu** [comment]
@alexey-milovidov Hi, CI checks had all passed, could you please assign a reviewer for this PR?

File: src/Storages/MergeTree/IMergeTreeDataPart.cpp [hunk 1]
```diff
@@ -1285,6 +1302,85 @@ void IMergeTreeDataPart::loadRowsCount()
...
+        size_t rows_read = reader->readRows(current_mark, total_mark, continue_reading, rows_to_read, result);
+        if (!rows_read)
+        {
+            LOG_WARNING(storage.log, "Part {} has lightweight delete, but _row_exists column not found", name);
+            existing_rows_count = rows_count;
+            return;
+        }
+
+        current_row += rows_read;
+        current_mark += (rows_to_read == rows_read);
```
Comment (yakov-olkhovskiy) [inline src/Storages/MergeTree/IMergeTreeDataPart.cpp:1365]: this doesn't look correct - `rows_read` apparently can come out less than `rows_to_read` - in this case `current_mark` will not be advanced - while it doesn't matter for other than first read (if `continue_reading` is true then `current_mark` doesn't matter), for the first read it does matter and if on the first read we get less than `rows_to_read` then `current_mark` doesn't advance then on the next iteration `continue_reading` will remain false and reading will restart from mark 0

File: src/Storages/MergeTree/IMergeTreeDataPart.cpp [hunk 2]
```diff
@@ -1285,6 +1302,85 @@ void IMergeTreeDataPart::loadRowsCount()
     }
 }
 
+void IMergeTreeDataPart::readExistingRowsCount() const
+{
+    if (!supportLightweightDeleteMutate() || !hasLightweightDelete() || !storage.getSettings()->exclude_deleted_rows_for_part_size_in_merge
+        || existing_rows_count < rows_count || !getMarksCount())
+        return;
+
+    std::lock_guard lock(existing_rows_count_mutex);
```
Comment (yakov-olkhovskiy) [inline src/Storages/MergeTree/IMergeTreeDataPart.cpp:1311]: do we need to protect this with mutex? I'm not quite sure
@KochetovNicolai could you please take a look?

File: src/Storages/MergeTree/MergeTreeSettings.h [hunk 3]
```diff
@@ -73,6 +73,7 @@ struct Settings;
     M(UInt64, number_of_mutations_to_throw, 1000, "If table has at least that many unfinished mutations, throw 'Too many mutations' exception. Disabled if set to 0", 0) \
     M(UInt64, min_delay_to_mutate_ms, 10, "Min delay of mutating MergeTree table in milliseconds, if there are a lot of unfinished mutations", 0) \
     M(UInt64, max_delay_to_mutate_ms, 1000, "Max delay of mutating MergeTree table in milliseconds, if there are a lot of unfinished mutations", 0) \
+    M(Bool, exclude_deleted_rows_for_part_size_in_merge, false, "If true, estimated size (excluding lightweight deleted rows) will be used as source part size when selecting parts to merge", 0) \
```
Comment (yakov-olkhovskiy) [inline src/Storages/MergeTree/MergeTreeSettings.h:76]:
```suggestion
    M(Bool, exclude_deleted_rows_for_part_size_in_merge, false, "Use an estimated part size (weighted by number of lightweight deleted rows) when selecting parts to merge", 0) \
```

File: src/Storages/MergeTree/IMergeTreeDataPart.cpp (hunk 1 above)
Comment (jewelzqiu) [reply src/Storages/MergeTree/IMergeTreeDataPart.cpp:1365]: I implemented this part with reference to [MergeTreeSequentialSource::generate()](https://github.com/ClickHouse/ClickHouse/blob/f503ea27e9a98d90c385ae2706ebfb5ca9a6315c/src/Storages/MergeTree/MergeTreeSequentialSource.cpp#L170C23-L185)
It looks like `reader->readRows()` will read all rows in this mark range (or up to specified `max_rows_to_read`)?

**Review by yakov-olkhovskiy — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
Sorry, this has been reverted because another reviewer (not me, I didn't read the code) has found an obvious race condition in the code. Let's fix this race condition and resubmit.

**Comment by alexey-milovidov** [comment]
Also, we have to add a test to make this race condition exposed.

**Comment by yakov-olkhovskiy** [comment]
@alexey-milovidov could you please specify what race condition we have here? do we have a note somewhere I missed?

**Comment by davenger** [comment]
Also, grabbing a mutex in `readExistingRowsCount()` and reading the whole `_row_exists` column from disk while holding it seems like a very dangerous approach. E.g. it can be called from `ReplicatedMergeTreeQueue::shouldExecuteLogEntry()` via `part->getExistingBytesOnDisk();` and can negatively affect replication queue processing.
I think the proper way is to save the count of "existing" rows in part's metadata  when part is created similar to total row count.

**Comment by yakov-olkhovskiy** [comment]
@davenger saving count is definitely better way of dealing with this but it requires metadata modification one way or another.
BTW do you know which race condition Alexey is talking about?

**Comment by jewelzqiu** [comment]
@davenger Actually my initial thought was to add an `existing_count.txt` (just like `count.txt`) before finishing mutation
As for existing parts, create `existing_count.txt` on loading process.
Do you think it is a better solution?

**Comment by jewelzqiu** [comment]
> @davenger Actually my initial thought was to add an `existing_count.txt` (just like `count.txt`) before finishing mutation As for existing parts, create `existing_count.txt` on loading process. Do you think it is a better solution?

@alexey-milovidov @yakov-olkhovskiy what are your thoughts about this approach?

**Comment by yakov-olkhovskiy** [comment]
I think it's better than lazy calculation, but backward compatibility should be taken into account.

**Comment by jewelzqiu** [comment]
> Sorry, this has been reverted because another reviewer (not me, I didn't read the code) has found an obvious race condition in the code. Let's fix this race condition and resubmit.

@alexey-milovidov Could you please provide more information about the race condition in this PR?

**Comment by alexey-milovidov** [comment]
I don't know what the race condition is, but my colleagues told me that it is there... I will clarify.

**Comment by jewelzqiu** [comment]
> I don't know what the race condition is, but my colleagues told me that it is there... I will clarify.

@alexey-milovidov thanks

**Comment by alexey-milovidov** [comment]
They said it was around `const bool & is_merge`

**Comment by jewelzqiu** [comment]
> They said it was around `const bool & is_merge`

I don't see any race conditions here... @yakov-olkhovskiy Do you have any clue?

**Comment by alexey-milovidov** [comment]
@jewelzqiu, My colleagues said it is related to `existing_rows_count`.

**Comment by jewelzqiu** [comment]
> @jewelzqiu, My colleagues said it is related to `existing_rows_count`.

@alexey-milovidov Thanks, I got it, looks like `existing_rows_count` can be read before lock acquired.
I have created another PR with different approach as discussed above in this PR: https://github.com/ClickHouse/ClickHouse/pull/58223
In the new PR `existing_rows_count` will only be set once before part became active
cc @yakov-olkhovskiy @davenger

**Pull request merged**